### ISSUE TITLE
`add` is too imprecise about video tracks

### ISSUE THREAD

**Issue by smimram** [savonet/liquidsoap#3209]
The current type of `add` is
```
(?id : string?, ?normalize : {bool}, ?power : {bool}, ?weights : [{float}],
 [source(audio=pcm('a), ?video=canvas('b), 'c)]) -> source(audio=pcm('a),
?video=canvas('b))
```
which is too imprecise about video tracks. When the input has video the output will too. We should find a way to express this.

For instance, the following simple script is not passing whereas it should:
```
s1 = single("video1.mp4")
s2 = single("video2.mkv")
s = add([s1, s2])
output.audio_video(s)
```
It raises
```
At ./test.liq, line 5, char 19:
output.audio_video(s)

Error 5: this value has type
  source(audio=_, ?video=canvas('A)) (inferred at source.liq, line 6 char 2 - line 39 char 5)
but it should be a subtype of the type of the value at io.liq, line 72, char 90
  source(audio=_, video='A)
```

**Comment by toots** [comment]
There are two ways to solve this for the release:
1. Add back hardcoded `add`
2. Add a specific `add.audio_video`.

I'm voting for 2 and I'm pretty sure you're voting for 1.

The reason I think 2 is a better option is because, as we introduce multitrack through the application, we do want the user to be aware of what's coming inside their sources.

I'm already not a fans of image sources being able to create implicit audio.

In general, I believe users manipulating audio or audio and video will have a very different mind set and, in this case, would not mind having to be specific about using a audio-only or audio+video operator. Not to mention cases with multiple audio tracks.

@Moonbase59 any opinion on this?

**Comment by smimram** [comment]
You're right I'm more in favor of 1 :)
- we should keep things as simple as possible
- if you want to add midi or some other track in the future will we have operators such as `add.video_and_midi_but_no_audio`? kind of heavy...

**Comment by Moonbase59** [comment]
Regarding the `add.video_and_midi_but_no_audio` example: No. Don’t make function naming too "hacky". why not have

- `add` – which would, per default, add everything that’s "addable", even creating new types of sources if needed, say, you "add" a video with no audio and an audio track → you get a video+audio source.
- `add(list_of_types_to_add, sources)` – in which you could set which types of tracks (video, audio, midi, subtitles) you wish to consider (if existant),
- or even expand this to define which actual tracks of the type(s) to add.

This is not thought out, just something how I would somehow expect "add" to work as an end user.

**Comment by smimram** [comment]
The hacky name was more intended as a joke :) My point is that if we can keep things simple and have only one `add` instead of many functions with clumsy naming conventions it would be better...

**Comment by Moonbase59** [comment]
> The hacky name was more intended as a joke :)

Sure :-)



> My point is that if we can keep things simple and have only one […]

Exactly my point. Better less and simple names, and more (optional) parameters.

**Comment by toots** [comment]
Yes, yes: Simple things should be simple but complex things should be possible has been our motto forever. I guess we want more simple. I'll change it back. @Moonbase59 this should also fix default transitions in crossfade with video.

**Comment by Moonbase59** [comment]
@toots Don’t get too impatient with me—was just trying to give feedback as an "average Joe" who is fighting since weeks to get something relatively simple done: Radio audio + video with some overlays… ;-)

Good news about the transitions!

Btw, any chance https://github.com/savonet/liquidsoap/issues/3202 gets fixed soon? I’m really stuck with that.
